### PR TITLE
Rewrite RGL to run graphs asynchronously

### PR DESCRIPTION
**Pull request by prybicki**

### CODE REVIEW

File: src/graph/GraphRunCtx.hpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		// exceptionPtr may be read by client's thread only after it acquire-read true `completed`
+		// exceptionPtr may be written by graph thread only before it release-stores true `completed`
+		std::exception_ptr exceptionPtr {nullptr};
+	};
+
+	std::unordered_map<Node::ConstPtr, NodeExecStatus> executionStatus;
+	std::atomic<bool> execThreadCanStart;
+
+	friend Node;
+	friend void handleDestructorException(std::exception_ptr e, const char* what);
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/GraphRunCtx.hpp:102]: Ok, applied everywhere.

File: src/graph/GraphRunCtx.cpp [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+#include <graph/GraphRunCtx.hpp>
+#include <graph/NodesCore.hpp>
+#include <graph/Node.hpp>
+#include <NvtxWrappers.hpp>
+
+std::list<std::shared_ptr<GraphRunCtx>> GraphRunCtx::instances;
+
+std::shared_ptr<GraphRunCtx> GraphRunCtx::createAndAttach(std::shared_ptr<Node> node)
+{
+	auto graphRunCtx = std::shared_ptr<GraphRunCtx>(new GraphRunCtx());
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/GraphRunCtx.cpp:24]: This is not possible. `GraphRunCtx()` constructor is private and cannot be called from within `std::make_shared`.

File: src/graph/GraphRunCtx.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+	graphRunCtx->nodes = node->getConnectedNodes();
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/GraphRunCtx.cpp:26]: If I understand correctly your question -  node is also connected to the graph context, just a few lines below:
```cpp
	for (auto&& currentNode : graphRunCtx->nodes) {
		if (currentNode->hasGraphRunCtx()) {
			auto msg = fmt::format("attempted to replace existing graphRunCtx in node {} when creating for {}", currentNode->getName(), node->getName());
			throw std::logic_error(msg);
		}
		currentNode->setGraphRunCtx(graphRunCtx);
	}
```

File: src/graph/GraphRunCtx.cpp [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+		status = cudaSuccess; // Ignore further checks.
+	}
+	CHECK_CUDA_NO_THROW(status);
+
+	if (maybeThread.has_value()) {
+		RGL_WARN("~GraphRunCtx(): maybeThread has value!");
+	}
+}
+
+void GraphRunCtx::detachAndDestroy()
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/GraphRunCtx.cpp:161]: This method is supposed to synchronize `GraphRunCtx` and remove all its references, which is expected to be followed by its destruction. After `detachAndDestroy`,  none of `this->nodes`  have `Node::graphRunCtx` set, so `Node::validate()` (ensuring `Node::graphRunCtx` is set) would fail with an exception.

File: src/graph/NodesCore.hpp [hunk 5]
```diff
@@ -338,22 +329,21 @@ struct TemporalMergePointsNode : Node, IPointsNodeSingleInput
...
+	{ return std::const_pointer_cast<const IAnyArray>(mergedData.at(field)); }
 
 private:
-	std::unordered_map<rgl_field_t, VArray::Ptr> mergedData;
+	std::unordered_map<rgl_field_t, IAnyArray::Ptr> mergedData;
 	std::size_t width = 0;
 };
 
-struct FromArrayPointsNode : Node, IPointsNode
+struct FromArrayPointsNode : virtual IPointsNode, virtual INoInputNode
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/NodesCore.hpp:340]: Thanks, fixed.

File: src/graph/Interfaces.hpp [hunk 6]
```diff
@@ -77,42 +89,52 @@ struct IPointsNode
...
 
-	Mat3x4f getLookAtOriginTransform() const override { return input->getLookAtOriginTransform(); }
+	virtual Mat3x4f getLookAtOriginTransform() const override { return input->getLookAtOriginTransform(); }
 
 	// Data getters
-	bool hasField(rgl_field_t field) const { return input->hasField(field); }
-	VArray::ConstPtr getFieldData(rgl_field_t field, cudaStream_t stream) const override
-	{ return input->getFieldData(field, stream); }
+	virtual bool hasField(rgl_field_t field) const { return input->hasField(field); }
+	virtual IAnyArray::ConstPtr getFieldData(rgl_field_t field) override
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/Interfaces.hpp:124]: Thanks for pointing out. These structs are going to be hopefully removed, so leaving as-is.

File: src/graph/Node.cpp [hunk 7]
```diff
@@ -13,27 +13,54 @@
...
+	if (child->hasGraphRunCtx()) {
+		child->getGraphRunCtx()->detachAndDestroy();
+	}
 
+	// Remove links
 	this->outputs.push_back(child);
 	child->inputs.push_back(shared_from_this());
+
+	// Sort outputs by priority
+	std::stable_sort(outputs.begin(), outputs.end(),
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/Node.cpp:53]: You are right about performance. Please let me also point out that for all real-world use-cases: `outputs.size() <= 3`

File: src/graph/Node.hpp [hunk 8]
```diff
@@ -16,55 +16,149 @@
...
+	 * @return True, if node can be executed.
+	 */
+	bool isValid() const { return !dirty; }
+
+	/**
+	 * Waits until execution of this node is completed (CPU & GPU).
+	 * This is required to query node for results.
+	 * If error happened in the graph thread, this function will re-throw it.
+	 */
+	void synchronize() const;
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/Node.hpp:147]: I agree with you, however `CompactPointsNode::getWidth() const` needs to synchronize itself before returning width, so it must call into `void synchronize() const`. This is clearly an architectural flaw. Leaving as-is for now.

File: src/graph/Node.cpp [hunk 9]
```diff
@@ -42,33 +69,154 @@ void Node::removeChild(Node::Ptr child)
...
-
-	// Remove us as a parent of that child
 	child->inputs.erase(thisIt);
+
+	this->dirty = true;
+	child->dirty = true;
 }
 
-std::shared_ptr<Graph> Node::getGraph()
+void Node::setGraphRunCtx(std::optional<std::shared_ptr<GraphRunCtx>> graph)
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/Node.cpp:102]: This friendship works the other way you suggested - Node can access `GraphRunCtx` private members. However, you suggestion makes sense, so I made the friendship bidirectional and made `Node::setGraphRunCtx` private.

File: src/api/apiCommon.hpp [hunk 10]
```diff
@@ -104,6 +108,47 @@ void createOrUpdateNode(rgl_node_t* nodeRawPtr, Args&&... args)
...
+	if (fieldsModified && node->hasGraphRunCtx()) {
+		node->getGraphRunCtx()->detachAndDestroy();
+	}
+
+	// As of now, there's no guarantee that changing node parameter won't influence other nodes
+	// Therefore, before changing them, we need to ensure all nodes are idle (not running in GraphRunCtx).
+	if (node->hasGraphRunCtx()) {
+		node->getGraphRunCtx()->synchronize();
+	}
 	node->setParameters(args...);
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/api/apiCommon.hpp:125]: Great catch. Some day it may be game-changing. Fixed.

File: src/graph/CompactPointsNode.cpp [hunk 11]
```diff
@@ -16,53 +16,69 @@
...
+		const auto* isHitPtr = input->getFieldDataTyped<IS_HIT_I32>()->asSubclass<DeviceAsyncArray>()->getReadPtr();
+		const CompactionIndexType * indices = inclusivePrefixSum->getReadPtr();
+		gpuApplyCompaction(getStreamHandle(), input->getPointCount(), getFieldSize(field), isHitPtr, indices, outPtr, inputPtr);
+		bool calledFromEnqueue = graphRunCtx.value()->isThisThreadGraphThread();
+		if (!calledFromEnqueue) {
+			// This is a special case, where API calls getFieldData for this field for the first time
+			// We did not enqueued compcation in enqueueExecImpl, yet, we are asked for results.
+			// This operation was enqueued in the graph stream, but API won't wait for whole graph stream.
+			// Therefore, we need a manual sync here.
+			// TODO: remove this cancer.
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/CompactPointsNode.cpp:71]: I see here excellent and adequate naming. As in "corruption is a cancer in our society", there's no vulgar/pejorative connotation and it correctly captures the intended meaning. I do really find this piece of code to be a cancer; something that should be removed, because its further growth may kill the entire project.

File: src/graph/DownSamplePointsNode.cpp [hunk 12]
```diff
@@ -42,66 +39,92 @@ void DownSamplePointsNode::schedule(cudaStream_t stream)
...
 
 	size_t offset = offsetof(PCLPoint, label);
 	size_t stride = sizeof(PCLPoint);
 	size_t size = sizeof(PCLPoint::label);
-	auto&& dst = (char*) filteredIndices->getDevicePtr();
-	auto&& src = (const char*) filteredPoints->getReadPtr(MemLoc::Device);
-	gpuCutField(stream, filtered->size(), dst, src, offset, stride, size);
-	CHECK_CUDA(cudaEventRecord(finishedEvent, stream));
+	auto&& dst = (char*) filteredIndices->getReadPtr();
+	auto&& src = (const char*) filteredPoints->getReadPtr();
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/DownSamplePointsNode.cpp:80]: Great catch. They actually were erroneous, because `filteredIndices->getReadPtr()` is incorrect here, should be `filteredIndices->getWritePtr()`.

File: src/graph/FromArrayPointsNode.cpp [hunk 13]
```diff
@@ -26,34 +27,29 @@ void FromArrayPointsNode::setParameters(const void* points, size_t pointCount, c
...
+	inputData->copyFromExternal(static_cast<const char *>(points), pointCount * pointSize);
+	const char* inputPtr = inputData->getReadPtr();
 
-void FromArrayPointsNode::validate()
-{
-	if (!inputs.empty()) {
-		auto msg = fmt::format("inputs for node {} are not allowed", getName());
-		throw InvalidPipeline(msg);
-	}
+	auto& gpuFields = gpuFieldDescBuilder.buildWritableAsync(arrayMgr.getStream(), getFieldToPointerMappings(fields));
```
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/FromArrayPointsNode.cpp:42]: If you don't mind, I'd prefer to change to `auto&&`.

File: test/include/helpers/lidarHelpers.hpp [hunk 14]
```diff
@@ -25,9 +25,30 @@ static std::vector<rgl_mat3x4f> makeLidar3dRays(float fov_x, float fov_y,
...
 }
+
+static std::vector<rgl_mat3x4f> makeGridOfParallelRays(Vec2f minCoord, Vec2f maxCoord, Vec2i count)
+{
+	std::vector<rgl_mat3x4f> rays;
+	EXPECT_TRUE(count.x() > 0);
+	EXPECT_TRUE(count.y() > 0);
+
+	float spacingY = (maxCoord.y() - minCoord.y()) / (count.y() + 1);
+	float spacingX = (maxCoord.x() - minCoord.x()) / (count.x() + 1);
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/include/helpers/lidarHelpers.hpp:42]: @nebraszka

File: test/include/helpers/textureHelpers.hpp [hunk 15]
```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <vector>
+#include <math.h>
+
+template<typename T>
+static std::vector<T>  generateStaticColorTexture(int width, int height, T value)
+{
+	auto texels = std::vector<T>(width * height);
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/include/helpers/textureHelpers.hpp:9]: @nebraszka

File: test/include/helpers/commonHelpers.hpp [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+#include <gtest/gtest.h>
+#include <gmock/gmock-matchers.h>
+
+#include <rgl/api/core.h>
+
+// TODO(msz-rai): Left this namespace for cleaner tests - fix namespace for rgl (Field)
+// using namespace ::testing;
+
+static constexpr float EPSILON_F = 1e-6f;
+static constexpr int maxGPUCoresTestCount = 20000;
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/include/helpers/commonHelpers.hpp:12]: @nebraszka

File: test/src/asyncStressTest.cpp [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+	if (firstN > 0) {
+		indices.resize(firstN);
+	}
+	return indices;
+}
+
+/**
+ * Helper structure that builds a graph with a minimal linear part and then a random subtree of TransformPointsNodes.
+ */
+struct RandomGraph
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/asyncStressTest.cpp:33]: :bow:

File: test/src/asyncStressTest.cpp [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+	}
+
+	// Function to re-randomize
+	void generateRandomRays()
+	{
+		// We want rays to hit the cube and do not bother with edge cases
+		std::uniform_real_distribution<float> randomFloat {-0.9, 0.9};
+		std::uniform_int_distribution<size_t> randomInt {1, 10};
+		raysXY = { randomFloat(randomGenerator), randomFloat(randomGenerator) };
+		rayCount = { randomInt(randomGenerator), randomInt(randomGenerator) };
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/asyncStressTest.cpp:94]: Initially, I thought you were right, but then re-read my code, and it turns out that it outsmarted us both.

I'm using `makeGridOfParallelRays` lazily here - notice that `raysXY` is passed as both `minCoord` and `maxCoord`, so all rays will be colinear, and raysXY are coordinates of this common line. Also, rayCount is a 2-tuple which describes how many rays will be on X and on Y, so it does not matter if `rayCount.x() > rayCount.y()`.

[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/asyncStressTest.cpp:94]: I skip changing this - in this test, it does not matter if the rays are colinear or not.

File: test/src/graph/nodes/FromArrayPointsNodeTest.cpp [hunk 19]
```diff
@@ -16,83 +22,46 @@ TEST_P(FromArrayPointsNodeTest, invalid_arguments)
...
         int pointsCount = GetParam();
-        inPoints = GenerateTestPointsArray(pointsCount);
-        usePointsNode = nullptr;
+		pointCloud = std::make_unique<TestPointCloud>(pointFields, pointsCount);
+		usePointsNode = nullptr;
     };
 
-    initializeArgumentsLambda();
-    EXPECT_RGL_INVALID_ARGUMENT(rgl_node_points_from_array(&usePointsNode, inPoints.data(), inPoints.size(), pointFields.data(), 0), "field_count > 0");
+	initializeArgumentsLambda();
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/graph/nodes/FromArrayPointsNodeTest.cpp:29]: Will be soon cleaned with the fire of clang-format ;)

File: test/src/helpers/pointsTest.cpp [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+		EXPECT_EQ(pointCloud.getFieldValues<DISTANCE_F32>(), distance);
+		EXPECT_EQ(pointCloud.getFieldValues<RETURN_TYPE_U8>(), returnType);
+		EXPECT_EQ(pointCloud.getFieldValues<TIME_STAMP_F64>(), timeStamp);
+	}
+};
+
+INSTANTIATE_TEST_SUITE_P(PointCloudTests, PointCloudTest, testing::Values(1, 100, maxGPUCoresTestCount),
+                         [](const auto& info) { return "pointsCount_" + std::to_string(info.param); });
+
+TEST_P(PointCloudTest, setter_and_getter_with_transform)
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/helpers/pointsTest.cpp:106]: @nebraszka

File: test/src/helpers/pointsTest.cpp [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+#include <helpers/commonHelpers.hpp>
+#include <helpers/mathHelpers.hpp>
+
+class PointCloudTest : public ::testing::TestWithParam<int>
+{
+protected:
+	std::vector<rgl_field_t> allNotDummyFields = {XYZ_F32,     IS_HIT_I32,  RAY_IDX_U32,  ENTITY_ID_I32,  INTENSITY_F32,
+	                                              RING_ID_U16, AZIMUTH_F32, DISTANCE_F32, RETURN_TYPE_U8, TIME_STAMP_F64};
+
+	std::vector<rgl_field_t> fieldsWithPaddings = {PADDING_32,   XYZ_F32,        PADDING_16,     IS_HIT_I32,  PADDING_8,
```
[the remark this answers is not in the document]
Comment (prybicki) [reply test/src/helpers/pointsTest.cpp:11]: Well, padding is supposed to create padding in binary output when using FormatPointsNode, so probably the tests are verifying if padding behavior is correct.

The more concerning thing here is violating `Single Source of Truth` by defining `allNotDummyFields`, delusionally hoping that it will be updated when new fields are added. Not gonna happen. I'd move it to `RGLFields.hpp.` @nebraszka

File: src/graph/GraphRunCtx.hpp [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+	 */
+	void synchronizeNodeCPU(Node::ConstPtr nodeToSynchronize);
+
+	bool isThisThreadGraphThread() const {return maybeThread.has_value() && maybeThread->get_id() == std::this_thread::get_id(); }
+
+	CudaStream::Ptr getStream() const { return stream; }
+	const std::set<std::shared_ptr<Node>>& getNodes() const { return nodes; }
+
+	virtual ~GraphRunCtx();
+private:
```
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply src/graph/GraphRunCtx.hpp]: Generally, I am against this, because it assumes/allows splitting class members of the same kind. Above, you have a constructor, a static method, some other method, few fields, then a static field, then nested class, and later there are again some fields. You are on point - this may improve overall understanding of class purposes, but IMO it lowers the readability. It may be be harder e.g. to see class members initialization order. Or to check, if some field is static or not.

This may be easier to handle via composition, if it is a matter of context (not only fields / methods split) - separate helper class responsible for some communication. On the other hand, fields / methods split is rather natural for most developers.

File: src/graph/GraphRunCtx.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply src/graph/GraphRunCtx.cpp:24]: Fair point, you are right!

File: src/graph/GraphRunCtx.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply src/graph/GraphRunCtx.cpp:26]: graphRunCts->nodes is set to node->getConnectedNodes - I assume that node itself is not between these nodes. Then, in for-loop you iterate over them, still not setting grapRunCtx for node that is a parameter of createAndAttach called here. Can you please show me the exact line, where this happens? :D Or, at least, just assure me that eveything is correct, if I am blind.

File: src/graph/GraphRunCtx.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply src/graph/GraphRunCtx.cpp:160]: Clear, thank you!

File: src/graph/Node.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply src/graph/Node.hpp:145]: Kay, lets leave it for now.

File: src/graph/CompactPointsNode.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply src/graph/CompactPointsNode.cpp:71]: |May it be :D

File: test/src/asyncStressTest.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply test/src/asyncStressTest.cpp]: Big brain time it was, fair point.

File: test/src/helpers/pointsTest.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply test/src/helpers/pointsTest.cpp:11]: On point, moving to RGLFields.hpp may be.

**Review by PawelLiberadzki — APPROVED** [review]

File: src/graph/GraphRunCtx.hpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/GraphRunCtx.hpp]: Ok, removed 2 redundant `private`, with only a comment left.

File: src/graph/GraphRunCtx.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/GraphRunCtx.cpp:26]: I meant `connected` as in `connected components`, which, by definition, include the node it is computed for. Renamed `getConnectedNodes` -> `getConnectedComponentNodes`.

[the remark this answers is not in the document]
Comment (prybicki) [reply src/graph/GraphRunCtx.cpp]: And here's the line where it happens:
image

[the remark this answers is not in the document]
Comment (PawelLiberadzki) [reply src/graph/GraphRunCtx.cpp]: Totally clear now, thank you!
